### ISSUE TITLE
Dice is 0 for images with no lesion

### ISSUE THREAD

**Issue by orouskhani** [MIC-DKFZ/nnUNet#1428]
Hello Fabian, 

Again, congrats on the new version of nnU-Net. 

I am working on the segmentation of brain aneurysms. Now, I have a question about the Dice coefficient calculation. 

This is the code for dice calculation provided by nnUnet:

*****************************************************************

    if confusion_matrix is None:
        confusion_matrix = ConfusionMatrix(test, reference)

    tp, fp, tn, fn = confusion_matrix.get_matrix()
    test_empty, test_full, reference_empty, reference_full = confusion_matrix.get_existence()

    if test_empty and reference_empty:
        if nan_for_nonexisting:
            return float("NaN")
        else:
            return 0.

    return float(2. * tp / (2 * tp + fp + fn))**

********************************************************************

Now, this is my question: 

1- For test data, for some images with no aneurysm, the model detects some false positive aneurysms. In such cases, the dice is 0, why? It seems that if the model detects 1 false positive or 1000 false positives, then the dice is 0. Is it reasonable?

2- In the code: 

> if test_empty and reference_empty:
>         if nan_for_nonexisting:
>             return float("NaN")
>         else:
>             return 0.

You have mentioned that the dice is "0" when the test and reference are empty. It means that when there is no aneurysm in reference and no aneurysm is detected by the model, the dice is "0". However, I believe in this case the dice should be "1". 

Thank you for your help.

**Comment by ykirchhoff** [comment]
Hi,

sorry for the late reply. You are correct that the dice is zero for empty ground truths regardless of whether there is 1 or 1000 false positives. This is simply a characteristic of the dice score, which only cares about relative overlap between ground truth and prediction. So no overlap -> dice=0.
If the predictions is also empty the dice is not strictly defined, so you might wanna go for nan (set `nan_for_nonexisting` to true) and treat these cases separately. If you want to assign a number, continuity would point to 0, a perfect segmentation to 1.
Long story short, dice is far from being a perfect metric and should always be handled with care ;)

I hope this helps.
Best,
Yannick
